Convert to use ed25519 keys / multihash.

dmitrizagidulin opened the pull request:
Based on Manu's work on did-veres-one (https://github.com/veres-one/did-veres-one/pull/6).

Closes #24 and #25.

(Depends on PR https://github.com/veres-one/veres-one-context/pull/3 ).

dmitrizagidulin committed:
Use v1 context from local npm package.

dmitrizagidulin committed:
Convert to use ed25519 keys / multihash.

dmitrizagidulin committed:
Generate DID URI based on capabilityInvocation key.

dmitrizagidulin committed:
Add ed25519 generate tests.

dmitrizagidulin committed:
Temporarily use veres-one-contract github branch (until PR published).

dmitrizagidulin commented:
Ready to review.

mattcollier commented on lib/methods/veres-one/veres-did-doc.js:
```diff
@@ -100,23 +104,31 @@ class VeresOneDidDoc {
     }
 
+    if(!keyPair) {
+      throw new Error('Cannot generate a cryptonym DID without a key')
```
missing semi.  There are a number of linting violations in the files in this repo.  Please lint with the latest rules.

dmitrizagidulin replied:
Thanks! Re-linted and fixed.

mattcollier commented on lib/methods/veres-one/veres-did-doc.js:
```diff
@@ -100,23 +104,31 @@ class VeresOneDidDoc {
-        'base58' : 'forge',
+      publicKey: keyPair.publicKey,
+      encoding: typeof keyPair.publicKey === 'string' ?
+        'ed25519' : 'forge',
```
one liner.

dmitrizagidulin replied:
Fixed, thanks.

mattcollier reviewed (changes requested):
I thought the aim was to have a did-veres-one dependency and use that code the generate the DID.  Is that correct?

dmitrizagidulin committed:
Minor linting related fixes.

dmitrizagidulin commented:
> I thought the aim was to have a did-veres-one dependency and use that code the generate the DID. Is that correct?

The plan is to eventually extract Veres-related code out of did-io into did-veres-one, but it is not a priority. All downstream libs should be using `did-io` meanwhile, anyway.

mattcollier committed:
Update deps.

mattcollier commented:
@dmitrizagidulin one of the tests is failing on node 8: https://travis-ci.org/digitalbazaar/did-io/jobs/435714189

dmitrizagidulin commented:
@mattcollier:

> one of the tests is failing on node 8: https://travis-ci.org/digitalbazaar/did-io/jobs/435714189

Yeah, that one's weird, it passes like 3/4 of the time, so something is off either with the algorithm or with the test. I need to check with @msporny or @dlongley

dmitrizagidulin committed:
Extract fingerprint() to LDKeyPair, impl. verifyFingerprint(), refactor tests.

dmitrizagidulin commented:
Ok, I extracted fingerprint generating code to `LDKeyPair` class, and added a corresponding `verifyFingerprint()` function. The reason the test was failing had to do with the vagaries of base58-encoding. When we compare bytes directly, the fingerprints verify.

mattcollier commented on lib/ld-key-pair.js:
```diff
@@ -207,6 +221,43 @@ class Ed25519KeyPair extends LDKeyPair {
+    const publicKeyBuffer = forge.util.binary.base58.decode(this.publicKey);
+
+    // slice off the two first multiformat bytes, 0x30 and 0xed
+    return publicKeyBuffer.equals(fingerprintBuffer.slice(2));
```
has the information being discarded here already been validated?  Does it need to be?

dmitrizagidulin replied:
Erm, fair point. Now validating.

dmitrizagidulin committed:
Validate the first two bytes of the fingerprint.

dlongley reviewed (approved):
LGTM
